pennylane_qulacs/qulacs_device.py

```diff
@@ -39,181 +74,239 @@
         [0, 0, 0, np.exp(1j * theta / 2)],
     ]
 )
-toffoli = np.diag([1 for i in range(8)])
-toffoli[6:8, 6:8] = np.array([[0, 1], [1, 0]])
 
 
-def hermitian(*args):
-    r"""Input validation for an arbitary Hermitian expectation.
+def _reverse_state(state_vector):
+    """Reverse the qubit order for a vector of amplitudes.
     Args:
-        args (array): square hermitian matrix
+        state_vector (iterable[complex]): vector containing the amplitudes
     Returns:
-        array: square hermitian matrix
+        list[complex]
     """
-    A = np.asarray(args[0])
-
-    if A.shape[0] != A.shape[1]:
-        raise ValueError('Expectation must be a square matrix.')
+    state_vector = np.array(state_vector)
+    N = int(np.log2(len(state_vector)))
+    reversed_state = state_vector.reshape([2] * N).T.flatten()
+    return list(reversed_state)
 
-    if not np.allclose(A, A.conj().T):
-        raise ValueError('Expectation must be Hermitian.')
 
-    return A
+# tolerance for numerical errors
+tolerance = 1e-10
 
 
-class QulacsDevice(Device):
+class QulacsDevice(QubitDevice):
     """Qulacs device"""
-    name = 'Qulacs device'
-    short_name = 'qulacs.simulator'
-    pennylane_requires = '>=0.5.0'
+    name = "Qulacs device"
+    short_name = "qulacs.simulator"
+    pennylane_requires = ">=0.10.0"
     version = __version__
-    author = 'Steven Oud and Xanadu'
+    author = "Steven Oud and Xanadu"
+    gpu_supported = GPU_SUPPORTED
 
     _capabilities = {
-        'model': 'qubit',
-        'tensor_observables': True
+        "model": "qubit",
+        "tensor_observables": True,
+        "inverse_operations": True
     }
 
-    _operations_map = {
-        'QubitStateVector': None,
-        'BasisState': None,
-        'QubitUnitary': None,
-        'Toffoli': toffoli,
-        'CSWAP': CSWAP,
-        'CRZ': crz,
-        'Rot': None,
-        'SWAP': gate.SWAP,
-        'CNOT': gate.CNOT,
-        'CZ': gate.CZ,
-        'S': gate.S,
-        'Sdg': gate.Sdag,
-        'T': gate.T,
-        'Tdg': gate.Tdag,
-        'RX': gate.RX,
-        'RY': gate.RY,
-        'RZ': gate.RZ,
-        'PauliX': gate.X,
-        'PauliY': gate.Y,
-        'PauliZ': gate.Z,
-        'Hadamard': gate.H
+    _operation_map = {
+        "QubitStateVector": None,
+        "BasisState": None,
+        "QubitUnitary": None,
+        "Toffoli": gate.TOFFOLI,
+        "CSWAP": gate.FREDKIN,
+        "CRZ": crz,
+        "SWAP": gate.SWAP,
+        "CNOT": gate.CNOT,
+        "CZ": gate.CZ,
+        "S": gate.S,
+        "T": gate.T,
+        "RX": gate.RX,
+        "RY": gate.RY,
+        "RZ": gate.RZ,
+        "Rot": None,
```

Comment:
If you want, you can remove the `Rot` gate from this dictionary, and in addition remove the implementation below within `apply()`.

If PennyLane detects that a device does not support `Rot`, it will decompose it automatically before passing it to the device. So if the user specifies `Rot`:

* PennyLane will first check if the device supports `Rot`. If so, it passes the rotation operation directly.

* If not, it will call `Rot.decomposition()`, and pass the result to the device (a list containing `[RZ, RY, RZ]`).

* Finally, if none of the _decomposed gates_ are supported, the user will finally get an error noting that `qulacs.simulator` does not support `qml.Rot`.

Reply:
Great, answers my question below! :)